### PR TITLE
[flutter_tool] New command project

### PR DESCRIPTION
**Pull request by Jasguerrero**
#2885 

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

### CODE REVIEW

File: packages/flutter_tools/lib/src/analyze_project_validator.dart [hunk 1]
```diff
@@ -0,0 +1,45 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'analyze_project.dart';
+import 'project.dart';
+
+abstract class ProjectValidatorTask {
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/analyze_project_validator.dart:8]: Nit: name this `ProjectValidator`. Note that this is more like a `DoctorValidator` than a `ValidatorTask` (which is a convenience class that wraps both a `DoctorValidator` and a `Future<ValidationResult>`.

```suggestion
abstract class ProjectValidator {
```

Comment (christopherfujino) [reply packages/flutter_tools/lib/src/analyze_project_validator.dart:8]: Also, I believe this can be made `const`.

File: packages/flutter_tools/lib/src/commands/analyze_project.dart [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+    return const FlutterCommandResult(ExitStatus.success);
+  }
+
+  void presentResults(final List<ProjectValidatorResult> results) {
+    StringBuffer buffer = StringBuffer();
+
+    for (ProjectValidatorResult result in results) {
+      addToBufferResult(result, buffer);
+      buffer.write('\n');
+      globals.logger.printBox(buffer.toString());
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/commands/analyze_project.dart:64]: inject vs. use `globals.`

File: packages/flutter_tools/lib/src/commands/analyze_project.dart [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+        break;
+      case Status.warning:
+        icon = '[!]';
+        break;
+      case Status.notReady:
+        icon = '[?]';
+        break;
+    }
+
+    buffer.write('$icon ${result.toString()}');
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/commands/analyze_project.dart:85]: nit. .toString() is unneeded, the default behavior of interpolating a non-string into a string is to call `.toString()` on it anyway

```suggestion
    buffer.write('$icon $result');
```

File: packages/flutter_tools/lib/src/commands/analyze_project.dart [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+
+  @override
+  final String description = 'Show information about the current project.';
+
+  @override
+  final String category = FlutterCommandCategory.project;
+
+  @override
+  Future<FlutterCommandResult> runCommand() async {
+    globals.flutterVersion.fetchTagsAndUpdate();
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/commands/analyze_project.dart:28]: why is this needed?

File: packages/flutter_tools/lib/src/commands/analyze_project.dart [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+
+    Directory workingDirectory;
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/commands/analyze_project.dart:30]: nit. prefer final and you can replace the if/else with a ternary

File: packages/flutter_tools/lib/src/commands/analyze_project.dart [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+        }
+      }
+    }
+
+    presentResults(results);
+    return const FlutterCommandResult(ExitStatus.success);
+  }
+
+  void presentResults(final List<ProjectValidatorResult> results) {
+    StringBuffer buffer = StringBuffer();
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/commands/analyze_project.dart:59]: this can be final

File: packages/flutter_tools/lib/src/commands/analyze_project.dart (hunk 2 above)
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/commands/analyze_project.dart:64]: Shouldn't you be printing the buffer outside of the for loop?

If this is a bug, can you write a test that would have caught it (for example, hard-code the expected logger output, so that we can see what it should look like)? Add it to packages/flutter_tools/test/commands.shard/hermetic/analyze_project_test.dart

**Review by christopherfujino — CHANGES_REQUESTED** [review]

File: packages/flutter_tools/lib/executable.dart [hunk 7]
```diff
@@ -146,6 +147,7 @@ List<FlutterCommand> generateCommands({
     terminal: globals.terminal,
     artifacts: globals.artifacts,
   ),
+  AnalyzeProjectCommand(verbose: verbose),
```
[the remark this answers is not in the document]
Comment (Jasguerrero) [reply packages/flutter_tools/lib/executable.dart:150]: `ValidateProjectCommand` SGTM

File: packages/flutter_tools/lib/src/analyze_project.dart [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+
+enum Status {
+  error,
+  warning,
+  success,
+  notReady,
+}
+
+class ProjectValidatorResult {
+  ProjectValidatorResult(this.name);
```
[the remark this answers is not in the document]
Comment (Jasguerrero) [reply packages/flutter_tools/lib/src/analyze_project.dart:15]: Sure.

[the remark this answers is not in the document]
Comment (Jasguerrero) [reply packages/flutter_tools/lib/src/analyze_project.dart:15]: I think a better way to tackle this is to make `ProjectValidatorResult` an abstract class and then having `ProjectValidatorResultSuccess` inherit from it.

File: packages/flutter_tools/lib/src/analyze_project_validator.dart [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+  List<SupportedPlatform> get supportedPlatforms;
+  List<ProjectValidatorResult> start(FlutterProject project);
```
[the remark this answers is not in the document]
Comment (christopherfujino) [reply packages/flutter_tools/lib/src/analyze_project_validator.dart:10]: Ok, this makes sense. For the doctor, we made a new abstraction, of a result that groups other results; however, over abstraction can be a thing :). Can you document this behavior in a dartdoc for `.start()`?

File: packages/flutter_tools/lib/src/analyze_project.dart [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+}
+
+class ProjectValidatorResult {
+
+  ProjectValidatorResult(this.name, this.value, this._status, {this.warning});
+
+  final String name;
+  final String value;
+  final String? warning;
+  final StatusProjectValidator _status;
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/analyze_project.dart:18]: might as well make this public and get rid of the getter.

File: packages/flutter_tools/lib/src/analyze_project.dart [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+
+enum StatusProjectValidator {
+  error,
+  warning,
+  success,
+}
+
+class ProjectValidatorResult {
+
+  ProjectValidatorResult(this.name, this.value, this._status, {this.warning});
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/analyze_project.dart:13]: nit. prefer named parameters when you have more than 2

File: packages/flutter_tools/lib/src/analyze_project_validator.dart [hunk 12]
```diff
@@ -0,0 +1,17 @@
...
+
+abstract class ProjectValidator {
+  bool supportsProject(FlutterProject project);
+
+  /// Can return more than one result in case a file/command have a lot of info to share to the user
+  Future<List<ProjectValidatorResult>> start(FlutterProject project);
+}
+
+const List <ProjectValidator> allProjectValidators = [
+  // add validators
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/analyze_project_validator.dart:16]:
```suggestion
  // TODO(jsguerrero): add validators
```

File: packages/flutter_tools/lib/src/analyze_project_validator.dart [hunk 13]
```diff
@@ -0,0 +1,17 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'analyze_project.dart';
+import 'project.dart';
+
+abstract class ProjectValidator {
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/analyze_project_validator.dart:8]: Can you add dartdoc's for all the new classes you add, and any non-obvious methods? (for example `.supportsProject` is self-explanatory)

File: packages/flutter_tools/lib/src/analyze_project_validator.dart [hunk 14]
```diff
@@ -0,0 +1,17 @@
...
+  bool supportsProject(FlutterProject project);
+
+  /// Can return more than one result in case a file/command have a lot of info to share to the user
+  Future<List<ProjectValidatorResult>> start(FlutterProject project);
+}
+
+const List <ProjectValidator> allProjectValidators = [
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/analyze_project_validator.dart:15]: nit. make this a static on the `ProjectValidator` class

File: packages/flutter_tools/lib/src/commands/analyze_project.dart [hunk 15]
```diff
@@ -0,0 +1,78 @@
...
+
+import '../analyze_project.dart';
+import '../analyze_project_validator.dart';
+import '../base/file_system.dart';
+import '../base/logger.dart';
+import '../project.dart';
+import '../runner/flutter_command.dart';
+
+class ValidateProjectCommand extends FlutterCommand {
+  ValidateProjectCommand(this.fileSystem, this.logger, {this.verbose = false});
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/commands/analyze_project.dart:13]: nit use named parameters when you have more than 2

File: packages/flutter_tools/lib/src/commands/analyze_project.dart [hunk 16]
```diff
@@ -0,0 +1,78 @@
...
+      buffer.write('\n');
+    }
+    logger.printBox(buffer.toString());
+  }
+
+  void addToBufferResult(ProjectValidatorResult result, StringBuffer buffer) {
+    String icon;
+    switch(result.status) {
+      case StatusProjectValidator.error:
+        icon = '[X]';
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/commands/analyze_project.dart:62]: We should be consistent with the doctor and use this unicode character: https://github.com/flutter/flutter/blob/master/packages/flutter_tools/lib/src/doctor_validator.dart#L258

File: packages/flutter_tools/lib/src/commands/analyze_project.dart [hunk 17]
```diff
@@ -0,0 +1,78 @@
...
+        break;
+      case StatusProjectValidator.success:
+        icon = '[✓]';
+        break;
+      case StatusProjectValidator.warning:
+        icon = '[!]';
+        break;
+    }
+
+    buffer.write('$icon $result');
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/commands/analyze_project.dart:72]: if you call `buffer.writeln()` here you don't need to print an extra newline in the above scope (line 53).
```suggestion
    buffer.writeln('$icon $result');
```

File: packages/flutter_tools/lib/src/analyze_project.dart (hunk 11 above)
Comment (christopherfujino) [reply packages/flutter_tools/lib/src/analyze_project.dart:13]: also, this constructor can be made `const`

File: packages/flutter_tools/lib/src/analyze_project.dart [hunk 18]
```diff
@@ -0,0 +1,37 @@
...
+  final String? warning;
+  final StatusProjectValidator _status;
+
+  StatusProjectValidator get status{
+    return _status;
+  }
+
+  @override
+  String toString() {
+    if (_status == StatusProjectValidator.error) {
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/analyze_project.dart:26]: if you are returning early, you can un-nest the conditionals to make reading this easier:

```dart
    if (_status == StatusProjectValidator.error) {
      return 'Error: $value';
    }
    String resultString = '$name: $value';
    if (warning != null) {
      return '$name: $value (warning: $warning)';
    }
    return '$name: $value';
  }
```

File: packages/flutter_tools/lib/src/commands/analyze_project.dart [hunk 19]
```diff
@@ -0,0 +1,78 @@
...
+    final String userPath = getUserPath();
+    final Directory workingDirectory = userPath.isEmpty ? fileSystem.currentDirectory : fileSystem.directory(userPath);
+
+    final FlutterProject project =  FlutterProject.fromDirectory(workingDirectory);
+    final List<ProjectValidatorResult> results = <ProjectValidatorResult>[];
+    final Set<ProjectValidator> ranValidators = <ProjectValidator>{};
+
+    for (final ProjectValidator validator in allProjectValidators) {
+      if (!ranValidators.contains(validator) && validator.supportsProject(project)) {
+        results.addAll(await validator.start(project));
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/commands/analyze_project.dart:39]: do we need to `await` one before starting the next, or are these safe to run concurrently?

Comment (christopherfujino) [inline packages/flutter_tools/lib/src/commands/analyze_project.dart:39]: Also, can we use a spinner like in the doctor, so that the user knows the tool is still running?

https://github.com/flutter/flutter/blob/master/packages/flutter_tools/lib/src/doctor.dart#L341

Comment (christopherfujino) [reply packages/flutter_tools/lib/src/commands/analyze_project.dart:39]: Also maybe we should catch exceptions and mark them as crashed, as we do in the tool: https://github.com/flutter/flutter/blob/master/packages/flutter_tools/lib/src/doctor.dart#L350

File: packages/flutter_tools/lib/src/commands/analyze_project.dart [hunk 20]
```diff
@@ -0,0 +1,78 @@
...
+
+    for (final ProjectValidatorResult result in results) {
+      addToBufferResult(result, buffer);
+      buffer.write('\n');
+    }
+    logger.printBox(buffer.toString());
+  }
+
+  void addToBufferResult(ProjectValidatorResult result, StringBuffer buffer) {
+    String icon;
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/commands/analyze_project.dart:59]: i think you can mark this as `final`
```suggestion
    final String icon;
```

File: packages/flutter_tools/test/commands.shard/hermetic/project_validator_test.dart [hunk 21]
```diff
@@ -0,0 +1,92 @@
...
+    return true;
+  }
+}
+
+void main() {
+  final BufferLogger loggerTest = BufferLogger.test();
+  FileSystem fileSystem;
+
+  group('analyze project command', () {
+    setUpAll(() {
```
Comment (christopherfujino) [inline packages/flutter_tools/test/commands.shard/hermetic/project_validator_test.dart:52]: Don't re-use the same filesystem between tests

```suggestion
    setUp(() {
```

File: packages/flutter_tools/test/commands.shard/hermetic/project_validator_test.dart [hunk 22]
```diff
@@ -0,0 +1,92 @@
...
+  }
+
+  @override
+  bool supportsProject(FlutterProject project) {
+    return true;
+  }
+}
+
+void main() {
+  final BufferLogger loggerTest = BufferLogger.test();
```
Comment (christopherfujino) [inline packages/flutter_tools/test/commands.shard/hermetic/project_validator_test.dart:48]: I'm thinking you probably shouldn't re-use the logger between tests?

File: packages/flutter_tools/test/commands.shard/hermetic/project_validator_test.dart [hunk 23]
```diff
@@ -0,0 +1,92 @@
...
+      final CommandRunner<void> runner = createTestCommandRunner(command);
+
+      await runner.run(<String>['validate-project']);
+
+      const String expected = '\n'
+          '┌──────────────────────────────────────────┐\n'
+          '│ [✓] pass: value                          │\n'
+          '│ [✗] Error: my error                      │\n'
+          '│ [!] pass two: pass (warning: my warning) │\n'
+          '│                                          │\n'
```
Comment (christopherfujino) [inline packages/flutter_tools/test/commands.shard/hermetic/project_validator_test.dart:70]: Can we get rid of this extra line?

File: packages/flutter_tools/lib/src/project_validator.dart [hunk 24]
```diff
@@ -0,0 +1,16 @@
...
+
+import 'project.dart';
+import 'project_validator_result.dart';
+
+abstract class ProjectValidator {
+  bool supportsProject(FlutterProject project);
+  /// Can return more than one result in case a file/command have a lot of info to share to the user
+  Future<List<ProjectValidatorResult>> start(FlutterProject project);
+  /// new ProjectValidators should be added here for the ValidateProjectCommand to run
+  static List <ProjectValidator> allProjectValidators = <ProjectValidator>[
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/project_validator.dart:13]:
```suggestion
  static const List <ProjectValidator> allProjectValidators = <ProjectValidator>[
```

File: packages/flutter_tools/test/commands.shard/hermetic/project_validator_test.dart [hunk 25]
```diff
@@ -0,0 +1,92 @@
...
+      final ValidateProjectCommand command = ValidateProjectCommand(
+          fileSystem: fileSystem,
+          logger: loggerTest,
+          allProjectValidators: <ProjectValidator>[ProjectValidatorDummy()]
+      );
+      final CommandRunner<void> runner = createTestCommandRunner(command);
+
+      await runner.run(<String>['validate-project']);
+
+      const String expected = '\n'
```
Comment (christopherfujino) [inline packages/flutter_tools/test/commands.shard/hermetic/project_validator_test.dart:65]: This test is great! Can you add a second validator, so we can see what it looks like?

**Review by christopherfujino — COMMENTED** [review]
This is coming along great. Just a few nits.

File: packages/flutter_tools/test/commands.shard/hermetic/project_validator_test.dart (hunk 25 above)
Comment (Jasguerrero) [reply packages/flutter_tools/test/commands.shard/hermetic/project_validator_test.dart:85]: @christopherfujino done

File: packages/flutter_tools/test/commands.shard/hermetic/project_validator_test.dart [hunk 26]
```diff
@@ -0,0 +1,113 @@
...
+            ProjectValidatorDummy(),
+            ProjectValidatorSecondDummy()
+          ]
+      );
+      final CommandRunner<void> runner = createTestCommandRunner(command);
+
+      await runner.run(<String>['validate-project']);
+
+      const String expected = '\n'
+          '┌──────────────────────────────────────────┐\n'
```
Comment (christopherfujino) [inline packages/flutter_tools/test/commands.shard/hermetic/project_validator_test.dart:86]: I think we should differentiate which of these came from which validator, the way we do with the doctor. If that information is not important, then we might as well flatten the structure and make each of these their own validator, and not have `start()` return multiple results.

Comment (Jasguerrero) [reply packages/flutter_tools/test/commands.shard/hermetic/project_validator_test.dart:86]: > I think we should differentiate which of these came from which validator, the way we do with the doctor. If that information is not important, then we might as well flatten the structure and make each of these their own validator, and not have `start()` return multiple results.

The main advantage of having more than one result returned was to avoid doing repetitive work on different validators (eg: parse a file more than once). Knowing the parent validator of each result can be beneficial for contributors trying to debug an error on a specific validator but I don't think It could be beneficial for the end user with the current implementation. I think a good fix would be adding a new field on the validators like `title` and first output that and then all of its results. eg:
`
class ProjectValidatorSecondDummy extends ProjectValidator  {
   final String title = 'Android'
}

class ProjectValidatorDummy extends ProjectValidator  {
   final String title = 'IOS tool'
}

//output
const String expected = '\n'
          '┌───────────────────┐\n'
            '│ Android                                                          │\n'
            '│ [✓] pass: value                                             │\n'
            '│ [✗] Error: my error                                       │\n'
            '│ [!] pass two: pass (warning: my warning) │\n'
            '│                                                                       │\n'
            '│ IOS tool                                                        │\n'
            '│ [✓] second: pass                                        │\n'
            '│ [✗] Error: second fail                                  │\n'
          '└──────────────────┘\n';
`

What do you think? @christopherfujino

Comment (christopherfujino) [reply packages/flutter_tools/test/commands.shard/hermetic/project_validator_test.dart:86]: This looks good to me.

Although to be consistent, I think we should always put the task name first (it looks like when there is an error, we print `"Error"` first).

Comment (Jasguerrero) [reply packages/flutter_tools/test/commands.shard/hermetic/project_validator_test.dart:95]: done @christopherfujino

File: packages/flutter_tools/lib/executable.dart [hunk 27]
```diff
@@ -146,6 +148,12 @@ List<FlutterCommand> generateCommands({
     terminal: globals.terminal,
     artifacts: globals.artifacts,
   ),
+  ValidateProjectCommand(
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/executable.dart:151]: I don't think we should add this until it's ready to be used.

**Review by christopherfujino — APPROVED** [review]
LGTM!
